scheduler/src/cook/mesos/rebalancer.clj

```diff
@@ -391,64 +391,66 @@
     (loop [state init-state
            remaining-preemption max-preemption
            [pending-job-ent & jobs-to-make-room-for] jobs-to-make-room-for
-           pending-job-ents-to-run []
-           task-ents-to-preempt []]
+           preemption-decisions []]
       (if (and pending-job-ent (pos? remaining-preemption))
         (let [[state' preemption-decision] (compute-next-state-and-preemption-decision offer-cache state params pending-job-ent)]
           (if preemption-decision
             (recur state'
                    (dec remaining-preemption)
                    jobs-to-make-room-for
-                   (conj pending-job-ents-to-run pending-job-ent)
-                   (into task-ents-to-preempt
-                         ;; If a task has no id, it must be a synthetic task.
-                         ;; This means that on one iteration of (compute-next-state-and-preemption-decision),
-                         ;; the rebalancer decided to make room for a certain hypothetical task,
-                         ;; but on a subsequent iteration, it became clear that OTHER hypothetical tasks would be an even better outcome.
-                         ;; We shouldn't try to actually preempt tasks that were never scheduled.
-                         (filter :db/id (:task preemption-decision))))
+                   (conj preemption-decisions (assoc preemption-decision
+                                                     :to-make-room-for pending-job-ent)))
             (recur state'
                    remaining-preemption
                    jobs-to-make-room-for
-                   pending-job-ents-to-run
-                   task-ents-to-preempt)))
-
+                   preemption-decisions)))
         (do
           (timers/stop timer)
-          (histograms/update! task-counts-to-preempt (count task-ents-to-preempt))
-          (histograms/update! job-counts-to-run (count pending-job-ents-to-run))
-          ;; pending-job-ents-to-run is only for debugging purpose
-          [pending-job-ents-to-run task-ents-to-preempt]
-          )))))
+          (histograms/update! task-counts-to-preempt (count (mapcat :task preemption-decisions)))
```

Comment:
I didn't realize reduce would be *that* much faster than mapcat so I reproduced your results:

```
cook.components=> (def a (repeatedly 1000 (fn [] {:tasks (repeat (rand-int 1000) :a)})))
#'cook.components/a
cook.components=> (quick-bench (count (mapcat :tasks a)))
Evaluation count : 18 in 6 samples of 3 calls.
             Execution time mean : 48.031183 ms
    Execution time std-deviation : 981.794831 µs
   Execution time lower quantile : 47.429016 ms ( 2.5%)
   Execution time upper quantile : 49.411953 ms (97.5%)
                   Overhead used : 1.719809 ns
nil
cook.components=> (quick-bench (reduce #(+ %1 (-> %2 :tasks count)) 0 a))
Evaluation count : 42 in 6 samples of 7 calls.
             Execution time mean : 15.862469 ms
    Execution time std-deviation : 172.828084 µs
   Execution time lower quantile : 15.693824 ms ( 2.5%)
   Execution time upper quantile : 16.061297 ms (97.5%)
                   Overhead used : 1.719809 ns
nil
```

However, even 20 tasks preempted for a single job would be quite large. Restricting to a 1000 preemptions, each which preempted 20 tasks, we see even better relative performance by reduce but both are fast enough that I'm leaning toward keeping as is. Is that ok with you?

```
cook.components=> (def b (repeatedly 1000 (fn [] {:tasks (repeat 20 :b)})))
#'cook.components/b
cook.components=> (quick-bench (count (mapcat :tasks b)))
Evaluation count : 294 in 6 samples of 49 calls.
             Execution time mean : 2.086402 ms
    Execution time std-deviation : 13.811516 µs
   Execution time lower quantile : 2.072167 ms ( 2.5%)
   Execution time upper quantile : 2.100173 ms (97.5%)
                   Overhead used : 1.719809 ns
nil
cook.components=> (quick-bench (reduce #(+ %1 (-> %2 :tasks count)) 0 b))
Evaluation count : 912 in 6 samples of 152 calls.
             Execution time mean : 660.952298 µs
    Execution time std-deviation : 4.295272 µs
   Execution time lower quantile : 657.737007 µs ( 2.5%)
   Execution time upper quantile : 667.002116 µs (97.5%)
                   Overhead used : 1.719809 ns
nil
```